tests/h/traversal/annotation_test.py

```diff
@@ -9,340 +8,109 @@
+        assert not permits(root, [], Permission.Annotation.CREATE)
 
-    def test_get_item_raises_when_annotation_is_not_found(
-        self, storage, pyramid_request
-    ):
-        factory = AnnotationRoot(pyramid_request)
-        storage.fetch_annotation.return_value = None
-
-        with pytest.raises(KeyError):
-            factory["123"]
-
-    def test_get_item_has_right_group_service(
-        self, pyramid_request, storage, groupfinder_service
+    def test_annotation_lookup(
+        self,
+        pyramid_request,
+        AnnotationContext,
+        storage,
+        groupfinder_service,
+        links_service,
     ):
-        factory = AnnotationRoot(pyramid_request)
-        storage.fetch_annotation.return_value = mock.Mock()
+        root = AnnotationRoot(pyramid_request)
 
-        resource = factory["123"]
-        assert resource.group_service == groupfinder_service
+        context = root[sentinel.annotation_id]
 
-    def test_get_item_has_right_links_service(
-        self, pyramid_request, storage, links_service
-    ):
-        factory = AnnotationRoot(pyramid_request)
-        storage.fetch_annotation.return_value = mock.Mock()
-
-        resource = factory["123"]
-        assert resource.links_service == links_service
+        assert context == AnnotationContext.return_value
+        storage.fetch_annotation.assert_called_once_with(
+            pyramid_request.db, sentinel.annotation_id
+        )
+        AnnotationContext.assert_called_once_with(
+            storage.fetch_annotation.return_value, groupfinder_service, links_service
+        )
 
     @pytest.fixture
     def storage(self, patch):
         return patch("h.traversal.annotation.storage")
 
+    @pytest.fixture
+    def AnnotationContext(self, patch):
+        return patch("h.traversal.annotation.AnnotationContext")
 
-@pytest.mark.usefixtures("groupfinder_service", "links_service")
-class TestAnnotationContext:
-    def test_links(self, groupfinder_service, links_service):
-        ann = mock.Mock()
-        res = AnnotationContext(ann, groupfinder_service, links_service)
 
-        result = res.links
+class TestAnnotationContext:
+    def test_links(self, annotation, context, links_service):
+        result = context.links
 
-        links_service.get_all.assert_called_once_with(ann)
+        links_service.get_all.assert_called_once_with(annotation)
         assert result == links_service.get_all.return_value
 
-    def test_link(self, groupfinder_service, links_service):
-        ann = mock.Mock()
-        res = AnnotationContext(ann, groupfinder_service, links_service)
-
-        result = res.link("json")
+    def test_link(self, annotation, context, links_service):
+        result = context.link("json")
 
-        links_service.get.assert_called_once_with(ann, "json")
+        links_service.get.assert_called_once_with(annotation, "json")
         assert result == links_service.get.return_value
```

Comment:
Instead of recreating a context object and making a fake annotation, this now uses a factory to create a real annotation and central fixture which bolts it together.

This makes it much clearer what is under test and means we don't have to import the services into every test.